src/frontend/query_commands.ml

```diff
@@ -359,6 +359,49 @@ let dispatch pipeline (type a) : a Query_protocol.t -> a =
     in
     List.map ~f:Mbrowse.node_loc path
 
+  | Type_definition pos ->
+    let typer = Mpipeline.typer_result pipeline in
+    let structures = Mbrowse.of_typedtree (Mtyper.get_typedtree typer) in
+    let pos = Mpipeline.get_lexing_pos pipeline pos in
+    let path = Mbrowse.enclosing pos [structures] in
+    let path =
+      let rec resolve_tlink env ty =
+        match ty.Types.desc with
+        | Tconstr (path, _, _) -> Some (env, path)
+        | Tlink ty -> resolve_tlink env ty
+        | _ -> None
+      in
+      Std.List.filter_map path ~f:(fun (env, node) ->
+        Locate.log ~title:"debug" "inspecting node: %s" (Browse_raw.string_of_node node);
+        match node with
+        | Browse_raw.Expression {exp_type = ty; _}
+        | Pattern {pat_type = ty; _}
+        | Core_type {ctyp_type = ty; _}
+        | Value_description { val_desc = { ctyp_type = ty; _ }; _ } ->
+          resolve_tlink env ty
+        | _ -> None
+      )
+    in
+    Std.List.filter_map path ~f:(fun (env, path) ->
+      Locate.log ~title:"debug" "found type: %s" (Path.name path);
+      let local_defs = Mtyper.get_typedtree typer in
+      match
+        Locate.from_string
+          ~config:(Mpipeline.final_config pipeline)
+          ~env ~local_defs ~pos ~namespaces:[`Type] `MLI
+          (* FIXME: instead of converting to a string, pass it directly. *)
+          (Path.name path)
```

Comment:
I agree with the FIXME: that's not how one should do things.
Instead of calling `Locate.from_string`, you'll want to expose (and rename?) `Locate.from_completion_entry` which works on a `Path.t` directly.

Reply:
I tried fixing this but it wasn't so simple. Will likely need some tweaks (or a rewrite?) from you.